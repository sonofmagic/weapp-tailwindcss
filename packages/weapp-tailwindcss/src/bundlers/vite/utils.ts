import type { ExistingRawSourceMap } from 'rollup'
import path from 'node:path'
import process from 'node:process'

export function slash(p: string): string {
  return p.replace(/\\/g, '/')
}

export const isWindows = process.platform === 'win32'

const cssLangs = `\\.(css|less|sass|scss|styl|stylus|pcss|postcss)($|\\?)`
export const cssLangRE = new RegExp(cssLangs)
export function isCSSRequest(request: string): boolean {
  return cssLangRE.test(request)
}

export function normalizePath(id: string): string {
  return path.posix.normalize(isWindows ? slash(id) : id)
}

const postfixRE = /[?#].*$/
export function cleanUrl(url: string): string {
  return url.replace(postfixRE, '')
}

export async function formatPostcssSourceMap(
  rawMap: ExistingRawSourceMap,
  file: string,
): Promise<ExistingRawSourceMap> {
  const inputFileDir = path.dirname(file)

  const sources = rawMap.sources.map((source) => {
    const cleanSource = cleanUrl(decodeURIComponent(source))

    // postcss virtual files
    if (cleanSource[0] === '<' && cleanSource.endsWith('>')) {
      return `\0${cleanSource}`
    }

    return normalizePath(path.resolve(inputFileDir, cleanSource))
  })

  return {
    file,
    mappings: rawMap.mappings,
    names: rawMap.names,
    sources,
    sourcesContent: rawMap.sourcesContent,
    version: rawMap.version,
  }
}
