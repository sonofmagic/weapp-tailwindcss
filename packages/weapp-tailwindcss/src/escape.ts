export * from '@weapp-core/escape'
export { jsStringEscape } from '@ast-core/escape'
