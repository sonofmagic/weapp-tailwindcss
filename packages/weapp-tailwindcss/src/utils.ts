import { createDefu } from 'defu'
// import type { Defu } from 'defu'
import type { InternalUserDefinedOptions } from '@/types'

export function isRegexp(value: unknown) {
  return Object.prototype.toString.call(value) === '[object RegExp]'
}

export function isMap(value: unknown) {
  return Object.prototype.toString.call(value) === '[object Map]'
}
export function regExpTest(arr: (string | RegExp)[] = [], str: string) {
  if (Array.isArray(arr)) {
    for (const item of arr) {
      if (typeof item === 'string') {
        if (str.includes(item)) {
          return true
        }
      }
      else if (isRegexp(item)) {
        item.lastIndex = 0
        if (item.test(str)) {
          return true
        }
      }
    }
    return false
  }
  throw new TypeError('paramater \'arr\' should be a Array of Regexp | String !')
}

export function noop() {}

// const MAX_ASCII_CHAR_CODE = 127

// export function isAscii(str: string) {
//   for (let i = 0, strLen = str.length; i < strLen; ++i) {
//     if (str.charCodeAt(i) > MAX_ASCII_CHAR_CODE) { return false }
//   }
//   return true
// }

export function groupBy<T>(arr: T[], cb: (arg: T) => string): Record<string, T[]> {
  if (!Array.isArray(arr)) {
    throw new TypeError('expected an array for first argument')
  }

  if (typeof cb !== 'function') {
    throw new TypeError('expected a function for second argument')
  }

  const result: Record<string, T[]> = {}
  for (const item of arr) {
    const bucketCategory = cb(item)
    const bucket = result[bucketCategory]

    if (Array.isArray(bucket)) {
      result[bucketCategory].push(item)
    }
    else {
      result[bucketCategory] = [item]
    }
  }

  return result
}

export function getGroupedEntries<T>(entries: [string, T][], options: InternalUserDefinedOptions) {
  const { cssMatcher, htmlMatcher, jsMatcher, wxsMatcher } = options
  const groupedEntries = groupBy(entries, ([file]) => {
    if (cssMatcher(file)) {
      return 'css'
    }
    else if (htmlMatcher(file)) {
      return 'html'
    }
    else if (jsMatcher(file) || wxsMatcher(file)) {
      return 'js'
    }
    else {
      return 'other'
    }
  })
  return groupedEntries as Record<'css' | 'html' | 'js' | 'other', [string, T][]>
}

export function removeExt(file: string) {
  return file.replace(/\.[^./]+$/, '')
}

export { default as defu } from 'defu'

export const defuOverrideArray = createDefu((obj, key, value) => {
  if (Array.isArray(obj[key]) && Array.isArray(value)) {
    obj[key] = value
    return true
  }
})
