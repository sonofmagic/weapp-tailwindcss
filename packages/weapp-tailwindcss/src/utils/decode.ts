const unicodeEscapeRE = /\\u([\dA-Fa-f]{4})/g

export function decodeUnicode(value: string) {
  return value.replace(unicodeEscapeRE, (_match, hex) => {
    const codePoint = Number.parseInt(hex, 16)
    return Number.isNaN(codePoint) ? _match : String.fromCharCode(codePoint)
  })
}

export function decodeUnicode2(input: string) {
  try {
    return JSON.parse(`"${input}"`)
  }
  catch (_error) {
    return decodeUnicode(input)
  }
}
