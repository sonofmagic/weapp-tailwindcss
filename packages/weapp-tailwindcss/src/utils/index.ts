import type { InternalUserDefinedOptions } from '../types'
import { defu, defuOverrideArray, groupBy, isMap, isRegexp, noop, regExpTest, removeExt } from '@weapp-tailwindcss/shared'

export {
  defu,
  defuOverrideArray,
  groupBy,
  isMap,
  isRegexp,
  noop,
  regExpTest,
  removeExt,
}

export type EntryGroup = 'css' | 'html' | 'js' | 'other'

function classifyEntry(filename: string, options: InternalUserDefinedOptions): EntryGroup {
  if (options.cssMatcher(filename)) {
    return 'css'
  }
  if (options.htmlMatcher(filename)) {
    return 'html'
  }
  if (options.jsMatcher(filename) || options.wxsMatcher(filename)) {
    return 'js'
  }
  return 'other'
}

function createEmptyGroups<T>(): Record<EntryGroup, [string, T][]> {
  return {
    css: [],
    html: [],
    js: [],
    other: [],
  }
}

export function getGroupedEntries<T>(entries: [string, T][], options: InternalUserDefinedOptions) {
  const groups = createEmptyGroups<T>()
  for (const entry of entries) {
    const [filename] = entry
    const group = classifyEntry(filename, options)
    groups[group].push(entry)
  }
  return groups
}

// const MAX_ASCII_CHAR_CODE = 127

// export function isAscii(str: string) {
//   for (let i = 0, strLen = str.length; i < strLen; ++i) {
//     if (str.charCodeAt(i) > MAX_ASCII_CHAR_CODE) { return false }
//   }
//   return true
// }
