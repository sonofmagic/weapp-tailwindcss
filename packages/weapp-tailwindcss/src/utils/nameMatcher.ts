export type NameMatcher = (value: string) => boolean

export function createNameMatcher(
  list: (string | RegExp)[] | undefined,
  { exact = false }: { exact?: boolean } = {},
): NameMatcher {
  if (!list || list.length === 0) {
    return () => false
  }

  const exactStrings = new Set<string>()
  const fuzzyStrings: string[] = []
  const regexList: RegExp[] = []

  for (const item of list) {
    if (typeof item === 'string') {
      if (exact) {
        exactStrings.add(item)
      }
      else {
        fuzzyStrings.push(item)
      }
    }
    else {
      regexList.push(item)
    }
  }

  return (value: string) => {
    if (exact && exactStrings.has(value)) {
      return true
    }
    if (!exact) {
      for (const candidate of fuzzyStrings) {
        if (value.includes(candidate)) {
          return true
        }
      }
    }
    for (const regex of regexList) {
      regex.lastIndex = 0
      if (regex.test(value)) {
        return true
      }
    }
    return false
  }
}
