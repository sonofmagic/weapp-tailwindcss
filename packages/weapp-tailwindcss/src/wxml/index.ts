export { replaceWxml } from './shared'
export { generateCode, templateReplacer, customTemplateHandler, createTemplateHandler, extractSource } from './utils'
