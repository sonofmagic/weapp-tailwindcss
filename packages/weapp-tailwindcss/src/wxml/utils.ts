import * as t from '@babel/types'
import { Parser } from 'htmlparser2'
import MagicString from 'magic-string'
import { replaceWxml } from './shared'
import { parseExpression, traverse } from '@/babel'
import type { ItemOrItemArray } from '@/reg'
import { variableRegExp } from '@/reg'
import { defuOverrideArray } from '@/utils'
import type { ITemplateHandlerOptions, RawSource } from '@/types'
import { replaceHandleValue } from '@/js/handlers'

// mpx class 中外部用单引号，所以 内部要用双引号
// 反之亦然
// function getQuotes(quote: string | null | undefined) {
//   return quote === "'" ? 'double' : 'single'
// }

export function generateCode(match: string, options: ITemplateHandlerOptions = {}) {
  try {
    const ms = new MagicString(match)
    const ast = parseExpression(match)

    traverse(ast, {
      StringLiteral(path) {
        // [g['人生']==='你好啊'?'highlight':'']
        if (t.isMemberExpression(path.parent)) {
          return
        }
        // parentPath maybe null
        // ['td',[(g.type==='你好啊')?'highlight':'']]
        if (t.isBinaryExpression(path.parent) && t.isConditionalExpression(path.parentPath?.parent)) {
          return
        }
        const n = path.node

        replaceHandleValue(
          n.value,
          n,
          {
            mangleContext: options.mangleContext,
            escapeMap: options.escapeMap,
            classNameSet: options.runtimeSet,
            needEscaped: true,
            always: true,
          },
          ms,
          1,
        )
        // path.node.value = replaceWxml(path.node.value, options)
      },
      noScope: true,
    })

    // const { code } = generate(ast, {
    //   compact: true,
    //   minified: true,
    //   jsescOption: {
    //     quotes: getQuotes(options.quote),
    //     minimal: true
    //   }
    // })

    return ms.toString()
  }
  catch {
    // https://github.com/sonofmagic/weapp-tailwindcss/issues/274
    // {{class}}
    return match
  }
}

/**
 * @internal
 */
function extract(original: string, reg: RegExp) {
  let match = reg.exec(original)
  const sources: RawSource[] = []

  while (match !== null) {
    // 过滤空字符串
    // if (match[1].trim().length) {
    const start = match.index
    const end = reg.lastIndex
    sources.push({
      start,
      end,
      raw: match[1],
    })

    match = reg.exec(original)
  }
  return sources
}

export function extractSource(original: string) {
  return extract(original, variableRegExp)
}

export function templateReplacer(original: string, options: ITemplateHandlerOptions = {}) {
  const sources = extractSource(original)

  if (sources.length > 0) {
    const resultArray: string[] = []
    let p = 0
    for (let i = 0; i < sources.length; i++) {
      const m = sources[i]
      const before = original.slice(p, m.start)

      // 匹配前值
      resultArray.push(
        replaceWxml(before, {
          keepEOL: true,
          escapeMap: options.escapeMap,
          mangleContext: options.mangleContext,
        }),
      )
      p = m.start
      // 匹配后值
      if (m.raw.trim().length > 0) {
        const code = generateCode(m.raw, options)
        const source = `{{${code}}}`
        m.source = source
      }
      else {
        m.source = ''
      }

      resultArray.push(m.source)
      p = m.end
      // 匹配最终尾部值
      if (i === sources.length - 1) {
        const after = original.slice(m.end)
        resultArray.push(
          replaceWxml(after, {
            keepEOL: true,
            escapeMap: options.escapeMap,
            mangleContext: options.mangleContext,
          }),
        )
      }
    }

    return resultArray.filter(Boolean).join('').trim()
  }
  else {
    return replaceWxml(original, {
      keepEOL: false,
      escapeMap: options.escapeMap,
      mangleContext: options.mangleContext,
    })
  }
}

function regTest(reg: RegExp, str: string) {
  reg.lastIndex = 0
  return reg.test(str)
}

export function isPropsMatch(props: ItemOrItemArray<string | RegExp>, attr: string) {
  if (Array.isArray(props)) {
    for (const prop of props) {
      const res = typeof prop === 'string' ? prop.toLowerCase() === attr.toLowerCase() : regTest(prop, attr)
      if (res) {
        return res
      }
    }
    return false
  }
  else if (typeof props === 'string') {
    return props === attr
  }
  else {
    return regTest(props, attr)
  }
}

export async function customTemplateHandler(rawSource: string, options: Required<ITemplateHandlerOptions>) {
  const {
    customAttributesEntities = [],
    disabledDefaultTemplateHandler,
    inlineWxs,
    runtimeSet,
    jsHandler,
  } = options ?? {}
  const s = new MagicString(rawSource)
  let tag = ''
  const wxsArray: {
    startIndex: number
    endIndex: number
    data: string
  }[] = []
  const parser = new Parser(
    {
      onopentagname(name) {
        tag = name
      },

      onattribute(name, value, quote) {
        if (value) {
          // https://github.com/fb55/htmlparser2/blob/5eea942451c1b836999d4557ed98470678c789b9/src/Parser.ts#L431
          function update() {
            s.update(
              parser.startIndex + name.length + 2,
              // !important
              // htmlparser2 9.0.0: parser.endIndex
              // htmlparser2 9.1.0: parser.endIndex - 1
              // https://github.com/sonofmagic/weapp-tailwindcss/issues/269
              parser.endIndex - 1,
              templateReplacer(value, {
                ...options,
                quote,
              }),
            )
          }
          // add 'virtualHostClass' toLowerCase
          if (
            !disabledDefaultTemplateHandler
            && (name === 'class' || name === 'hover-class' || name === 'virtualHostClass' || name === 'virtualhostclass')
          ) {
            update()
          }
          for (const [t, props] of customAttributesEntities) {
            if (t === '*') {
              if (isPropsMatch(props, name)) {
                update()
              }
            }
            else if (typeof t === 'string') {
              if (t === tag && isPropsMatch(props, name)) {
                update()
              }
            }
            else if (regTest(t, tag) && isPropsMatch(props, name)) {
              update()
            }
          }
        }
      },
      ontext(data) {
        if (inlineWxs && tag === 'wxs') {
          wxsArray.push({
            data,
            endIndex: parser.endIndex + 1,
            startIndex: parser.startIndex,
          })
        }
      },
      onclosetag() {
        tag = ''
      },
    },
    {
      xmlMode: true,
    },
  )
  parser.write(s.original)
  parser.end()
  for (const { data, endIndex, startIndex } of wxsArray) {
    const { code } = await jsHandler(data, runtimeSet)
    s.update(startIndex, endIndex, code)
  }

  return s.toString()
}

export function createTemplateHandler(options: Omit<ITemplateHandlerOptions, 'runtimeSet'> = {}) {
  return (rawSource: string, opt: Pick<ITemplateHandlerOptions, 'runtimeSet'> = {}) => {
    return customTemplateHandler(rawSource, defuOverrideArray(opt, options) as Required<ITemplateHandlerOptions>)
  }
}
