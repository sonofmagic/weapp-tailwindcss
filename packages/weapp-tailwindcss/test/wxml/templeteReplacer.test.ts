import { templateReplacer } from '@/wxml/utils'
import { MappingChars2String, SimpleMappingChars2String } from '@/escape'

const testTable = [[{}]]

function complexReplacer(str: string) {
  return templateReplacer(str, {
    escapeMap: MappingChars2String,
  })
}
function simpleReplacer(str: string) {
  return templateReplacer(str, {
    escapeMap: SimpleMappingChars2String,
  })
}

describe('templateReplacer', () => {
  it.each(testTable)('$label isStringLiteral', () => {
    const testCase = '{{[\'som-node__label\',\'data-v-59229c4a\',\'som-org__text-\'+(node.align||\'\'),node.active||collapsed?\'som-node__label-active\':\'\',d]}}'

    const result = complexReplacer(testCase)

    expect(result).toBe('{{[\'som-node__label\',\'data-v-59229c4a\',\'som-org__text-\'+(node.align||\'\'),node.active||collapsed?\'som-node__label-active\':\'\',d]}}')
    expect(result).toMatchSnapshot()
  })

  it('isConditionalExpression', () => {
    const testCase = '{{[\'flex\',\'flex-col\',\'items-center\',flag===1?\'bg-red-900\':\'bg-[#fafa00]\']}}'
    const result = complexReplacer(testCase)
    expect(result).toBe('{{[\'flex\',\'flex-col\',\'items-center\',flag===1?\'bg-red-900\':\'bg-_bl__h_fafa00_br_\']}}')
    expect(result).toMatchSnapshot()
  })

  it('nest ', () => {
    const testCase
      = '{{[flag?\'bg-red-900\':\'bg-[#fafa00]\',classObject,[(flag===true)?\'bg-[#fafa00]\':\'\',(true)?\'text-sm\':\'\'],flag?flag===false?\'bg-red-900\':\'bg-[#000]\':\'bg-[#fafa00]\']}}'

    const result = complexReplacer(testCase)
    expect(result).toBe(
      '{{[flag?\'bg-red-900\':\'bg-_bl__h_fafa00_br_\',classObject,[(flag===true)?\'bg-_bl__h_fafa00_br_\':\'\',(true)?\'text-sm\':\'\'],flag?flag===false?\'bg-red-900\':\'bg-_bl__h_000_br_\':\'bg-_bl__h_fafa00_br_\']}}',
    )
    expect(result).toMatchSnapshot()
  })

  it('sm:text-3xl dark:text-sky-400', () => {
    const testCase = 'sm:text-3xl dark:text-slate-200 bg-[#ffffaa]'
    const result = complexReplacer(testCase)
    expect(result).toBe('sm_c_text-3xl dark_c_text-slate-200 bg-_bl__h_ffffaa_br_')
    expect(result).toMatchSnapshot()
  })

  it('\\r\\n replace test', () => {
    const testCase = `
    bg-white
    rounded-full
    w-10
    h-10
    flex
    justify-center
    items-center
    pointer-events-auto
  `
    const result = complexReplacer(testCase)
    expect(result).toBe('    bg-white    rounded-full    w-10    h-10    flex    justify-center    items-center    pointer-events-auto  ')
  })

  it('\\r\\n replace test with var', () => {
    const testCase = `{{[
      'flex',
      'items-center',
      'justify-center',
      'h-[100px]',
      'w-[100px]',
      'rounded-[40px]',
      'bg-[#123456]',
      'bg-opacity-[0.54]',
      'text-[#ffffff]',
      'data-v-1badc801',
      'text-[#123456]',
      b]}}`
    const result = simpleReplacer(testCase)
    expect(result).toBe(
      `{{[
      'flex',
      'items-center',
      'justify-center',
      'h-_100px_',
      'w-_100px_',
      'rounded-_40px_',
      'bg-_h123456_',
      'bg-opacity-_0d54_',
      'text-_hffffff_',
      'data-v-1badc801',
      'text-_h123456_',
      b]}}`,
    )
  })

  it('variables with multiple literal', () => {
    const testCase = `border-0 icon h-10 w-10 mx-auto {{active=='home'? 'icon-home-selected' : 'icon-home'}} {{}} {{ }} w-[20px] {{flag=='p-[20px]'? 'p-[20px]' : 'm-[20px]'}} h-[20px]`
    const result = complexReplacer(testCase)
    expect(result).toBe(
      'border-0 icon h-10 w-10 mx-auto {{active==\'home\'? \'icon-home-selected\' : \'icon-home\'}} {{}} {{ }} w-_bl_20px_br_ {{flag==\'p-[20px]\'? \'p-_bl_20px_br_\' : \'m-_bl_20px_br_\'}} h-_bl_20px_br_',
    )
  })

  it.each(testTable)('variables with multiple literal(2)', () => {
    const testCase = `border-0 icon h-10 w-10 mx-auto {{active=='home'? 'icon-home-selected' : 'icon-home'}} {{b}} {{ a==='cc' }} w-[20px] {{flag=='p-[20px]'? 'p-[20px]' : 'm-[20px]'}}`
    const result = complexReplacer(testCase)
    expect(result).toBe(
      'border-0 icon h-10 w-10 mx-auto {{active==\'home\'? \'icon-home-selected\' : \'icon-home\'}} {{b}} {{ a===\'cc\' }} w-_bl_20px_br_ {{flag==\'p-[20px]\'? \'p-_bl_20px_br_\' : \'m-_bl_20px_br_\'}}',
    )
  })

  it.each(testTable)('%label for toutiao str add not array', () => {
    const testCase = '{{(\'!font-bold\') + \' \' + \'!text-[#990000]\' + \' \' + \'data-v-1badc801\' + \' \' + \'text-2xl\' + \' \' + b}}' // '{{\'font-bold\'+\'\'+\'text-blue-500\'+\'\'+\'data-v-1badc801\'+\'\'+\'text-2xl\'+\'\'+b}}'

    const result = complexReplacer(testCase)
    expect(result).toBe('{{(\'_i_font-bold\') + \' \' + \'_i_text-_bl__h_990000_br_\' + \' \' + \'data-v-1badc801\' + \' \' + \'text-2xl\' + \' \' + b}}')
  })

  it.each(testTable)('%label utils.bem()', () => {
    const testCase
      = 'custom-class {{ utils.bem(\'button\', [type, size, { block, round, plain, square, loading, disabled, hairline, unclickable: disabled || loading }]) }} {{ hairline ? \'van-hairline--surround\' : \'\' }}'

    const result = complexReplacer(testCase)
    expect(result).toBe(
      'custom-class {{ utils.bem(\'button\', [type, size, { block, round, plain, square, loading, disabled, hairline, unclickable: disabled || loading }]) }} {{ hairline ? \'van-hairline--surround\' : \'\' }}',
    )
  })

  // it.each(testTable)('%label class with string var', ({ mangle }) => {
  //   const testCase = 'btn a{{num}}'
  //   const str = complexReplacer(testCase, { classGenerator: mangle ? classGenerator : undefined })
  //   expect(str).toBe(testCase)
  // })

  it('classGenerator class with string var', () => {
    const testCase = 'btn-%1 a[p-1]{{num}}'
    // classGenerator
    const str = complexReplacer(testCase)
    expect(str).toBe('btn-_p_1 a_bl_p-1_br_{{num}}')
  })

  it('classGenerator class with string var case 0', () => {
    const testCase = 'btn-%1 a[p-1]{{num}}'
    // classGenerator
    const str = simpleReplacer(testCase)
    expect(str).toBe('btn-p1 a_p-1_{{num}}')
  })

  it('classGenerator class with string var case 1', () => {
    const testCase = 'btn-%1 a[p-1]{{num}}b[b-2]'
    // classGenerator
    const str = simpleReplacer(testCase)
    expect(str).toBe('btn-p1 a_p-1_{{num}}b_b-2_')
  })

  it('classGenerator class with string var case 2', () => {
    const testCase = 'a[p-1]{{num}}b[b-2]{{p}}'
    // classGenerator
    const str = simpleReplacer(testCase)
    expect(str).toBe('a_p-1_{{num}}b_b-2_{{p}}')
  })

  // .shadow-\[0px_2px_11px_0px_rgba\(0\2c 0\2c 0\2c 0\.4\)\]
  it('arbitrary shadow values 0', () => {
    // 逗号 comma 的原因
    const testCase = 'shadow-[0px_2px_11px_0px_rgba(0,0,0,0.4)]'

    const result = complexReplacer(testCase)
    expect(result).toBe('shadow-_bl_0px_2px_11px_0px_rgba_pl_0_co_0_co_0_co_0_d_4_qr__br_')
  })

  // .shadow-\[0px_2px_11px_0px_\#0000000a\]
  it('arbitrary shadow values 1', () => {
    const testCase = 'shadow-[0px_2px_11px_0px_#00000a]'
    const result = complexReplacer(testCase)
    expect(result).toBe('shadow-_bl_0px_2px_11px_0px__h_00000a_br_')
  })

  it('arbitrary before:content-[\'hello\']', () => {
    const testCase = 'before:content-[\'hello\']'
    const result = complexReplacer(testCase)
    expect(result).toBe('before_c_content-_bl__q_hello_q__br_')
  })

  it('two ConditionalExpression', () => {
    const testCase = 'btn a{{num >=\'p-[1]\'?num===\'q-[2]\'?\'x-[0]\':\'y-[1]\':\'z-[2]\'}}'
    const result = complexReplacer(testCase)
    expect(result).toBe('btn a{{num >=\'p-[1]\'?num===\'q-[2]\'?\'x-_bl_0_br_\':\'y-_bl_1_br_\':\'z-_bl_2_br_\'}}')
  })

  it.skip('start up with num case', () => {
    const nums = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

    for (const num of nums) {
      const testCase = `${num}xl:text-base`
      const result = templateReplacer(testCase)
      expect(result).toBe(`_${num}xlctext-base`)

      const netestCase = `-${num}xl:text-base`
      const neresult = templateReplacer(netestCase)
      expect(neresult).toBe(`_-${num}xlctext-base`)
    }
  })
  // https://www.w3.org/TR/css-syntax-3/#ident-token-diagram
  it.skip('only - escape', () => {
    let result = templateReplacer('-')
    expect(result).toBe(`_-`)
    result = templateReplacer('--')
    expect(result).toBe(`--`)
  })

  it('font-size#setting-the-line-height', () => {
    let testCase = 'text-sm/[17px]'

    expect(simpleReplacer(testCase)).toBe('text-sms_17px_')
    testCase = 'text-base/loose'
    expect(simpleReplacer(testCase)).toBe('text-basesloose')
    testCase = 'text-[64rpx]/[72rpx]'
    expect(simpleReplacer(testCase)).toBe('text-_64rpx_s_72rpx_')
  })

  it('issues/274 {{class}}', () => {
    expect(simpleReplacer(`{{class}}`)).toBe('{{class}}')
  })

  it('issues/276 case 0', () => {
    expect(simpleReplacer(`relative h-12 w-12 before:absolute before:inset-0 before:border-2 before:border-red-500 rounded-[20rpx] before:rounded-[20rpx]`)).toBe(
      'relative h-12 w-12 beforecabsolute beforecinset-0 beforecborder-2 beforecborder-red-500 rounded-_20rpx_ beforecrounded-_20rpx_',
    )
  })

  it('issues/276 case 1', () => {
    expect(simpleReplacer(`relative h-12 w-12 before:absolute before:inset-0 before:border-2 before:border-red-500 before:rounded-[20rpx] rounded-[20rpx]`)).toBe(
      'relative h-12 w-12 beforecabsolute beforecinset-0 beforecborder-2 beforecborder-red-500 beforecrounded-_20rpx_ rounded-_20rpx_',
    )
  })
  // https://github.com/sonofmagic/weapp-tailwindcss/issues/316
  it('issues/316 case 0', () => {
    expect(simpleReplacer(`{{tabActive === '1' && 'book-txt__active'}}`)).toBe(
      `{{tabActive === '1' && 'book-txt__active'}}`,
    )
  })

  it('rd-tag case 0', () => {
    expect(simpleReplacer(`rd-tag-{{type}}-{{theme}}`)).toBe(
      `rd-tag-{{type}}-{{theme}}`,
    )
  })

  it('isConditionalExpression case 0', () => {
    const testCase = '{{[\'flex\',\'flex-col\',\'items-center\',flag===1?\'bg-red-900\':\'bg-[#fafa00]\']}}'
    const result = templateReplacer(testCase)
    expect(result).toBe('{{[\'flex\',\'flex-col\',\'items-center\',flag===1?\'bg-red-900\':\'bg-_hfafa00_\']}}')
  })

  // it('handleEachClassFragment case 0', () => {
  //   expect(handleEachClassFragment(`rd-tag-{{type}}-{{theme}}`)).toBe(
  //     `rd-tag-{{type}}-{{theme}}`,
  //   )
  // })

  // it('handleEachClassFragment case 1', () => {
  //   expect(handleEachClassFragment(`{{type}}-{{theme}}`)).toBe(
  //     `{{type}}-{{theme}}`,
  //   )
  // })
})
