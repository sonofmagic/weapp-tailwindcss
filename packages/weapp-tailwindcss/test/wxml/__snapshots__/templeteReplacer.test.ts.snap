// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`templateReplacer > isConditionalExpression 1`] = `"{{['flex','flex-col','items-center',flag===1?'bg-red-900':'bg-_bl__h_fafa00_br_']}}"`;

exports[`templateReplacer > nest  1`] = `"{{[flag?'bg-red-900':'bg-_bl__h_fafa00_br_',classObject,[(flag===true)?'bg-_bl__h_fafa00_br_':'',(true)?'text-sm':''],flag?flag===false?'bg-red-900':'bg-_bl__h_000_br_':'bg-_bl__h_fafa00_br_']}}"`;

exports[`templateReplacer > sm:text-3xl dark:text-sky-400 1`] = `"sm_c_text-3xl dark_c_text-slate-200 bg-_bl__h_ffffaa_br_"`;

exports[`templateReplacer > undefined isStringLiteral 1`] = `"{{['som-node__label','data-v-59229c4a','som-org__text-'+(node.align||''),node.active||collapsed?'som-node__label-active':'',d]}}"`;
